File: WordFrequencyAnalyzer.java
import java.io.*;
import java.util.*;
import java.util.regex.Pattern;
import java.util.regex.Matcher;

/**
 * Word Frequency Analyzer
 * This program reads a text file and counts the frequency of each word.
 */
public class WordFrequencyAnalyzer {
    
    /**
     * Main entry point of the program
     */
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        
        System.out.print("Enter the path to the text file (or press Enter for 'sample_log.txt'): ");
        String filePath = scanner.nextLine().trim();
        
        if (filePath.isEmpty()) {
            filePath = "sample_log.txt";
        }
        
        System.out.print("Enter number of top words to display (default 10): ");
        String topNInput = scanner.nextLine().trim();
        int topN = 10;
        
        if (!topNInput.isEmpty()) {
            try {
                topN = Integer.parseInt(topNInput);
            } catch (NumberFormatException e) {
                System.out.println("Invalid input. Using default value of 10.");
            }
        }
        
        scanner.close();
        
        // Perform the analysis
        analyzeLogFile(filePath, topN);
    }
    
    /**
     * Reads a text file and returns its contents as a string.
     * 
     * @param filePath Path to the text file
     * @return Contents of the file, or null if an error occurs
     */
    public static String readTextFile(String filePath) {
        try {
            StringBuilder content = new StringBuilder();
            BufferedReader reader = new BufferedReader(new FileReader(filePath));
            
            String line;
            while ((line = reader.readLine()) != null) {
                content.append(line).append("\n");
            }
            
            reader.close();
            return content.toString();
        } catch (FileNotFoundException e) {
            System.out.println("Error: File '" + filePath + "' not found.");
            return null;
        } catch (IOException e) {
            System.out.println("Error reading file: " + e.getMessage());
            return null;
        }
    }
    
    /**
     * Extracts words from text, converting to lowercase and removing punctuation.
     * Uses regex to find word boundaries.
     * 
     * @param text Input text string
     * @return List of extracted words
     */
    public static List<String> extractWords(String text) {
        List<String> words = new ArrayList<>();
        
        // Regex pattern to match word boundaries: \b\w+\b
        // \b = word boundary, \w+ = one or more word characters
        Pattern pattern = Pattern.compile("\\b\\w+\\b");
        Matcher matcher = pattern.matcher(text.toLowerCase());
        
        while (matcher.find()) {
            words.add(matcher.group());
        }
        
        return words;
    }
    
    /**
     * Counts the frequency of each word using a HashMap.
     * 
     * @param words List of words to count
     * @return HashMap with word frequencies
     */
    public static Map<String, Integer> countWordFrequency(List<String> words) {
        Map<String, Integer> wordCount = new HashMap<>();
        
        // Count occurrences of each word
        for (String word : words) {
            wordCount.put(word, wordCount.getOrDefault(word, 0) + 1);
        }
        
        return wordCount;
    }
    
    /**
     * Sorts the word frequency map by value (frequency) in descending order
     * and returns the top N entries.
     * 
     * @param wordCount Map of word frequencies
     * @param topN Number of top words to return
     * @return List of Map.Entry sorted by frequency (descending)
     */
    public static List<Map.Entry<String, Integer>> getTopWords(
            Map<String, Integer> wordCount, int topN) {
        
        // Convert map entries to list
        List<Map.Entry<String, Integer>> entries = new ArrayList<>(wordCount.entrySet());
        
        // Sort by value (frequency) in descending order
        entries.sort((a, b) -> b.getValue().compareTo(a.getValue()));
        
        // Return top N entries
        int size = Math.min(topN, entries.size());
        return entries.subList(0, size);
    }
    
    /**
     * Displays the word frequency results in a formatted way.
     * 
     * @param wordCount Map of word frequencies
     * @param topN Number of top words to display
     */
    public static void displayResults(Map<String, Integer> wordCount, int topN) {
        System.out.println("\n" + "=".repeat(50));
        System.out.println("Total unique words: " + wordCount.size());
        System.out.println("=".repeat(50) + "\n");
        
        System.out.println("Top " + topN + " most frequent words:");
        System.out.println("-".repeat(50));
        System.out.printf("%-20s %-10s%n", "Word", "Frequency");
        System.out.println("-".repeat(50));
        
        // Get and display top N words
        List<Map.Entry<String, Integer>> topWords = getTopWords(wordCount, topN);
        for (Map.Entry<String, Integer> entry : topWords) {
            System.out.printf("%-20s %-10d%n", entry.getKey(), entry.getValue());
        }
        
        System.out.println("-".repeat(50) + "\n");
    }
    
    /**
     * Main function that orchestrates the word frequency analysis.
     * 
     * @param filePath Path to the text/log file
     * @param topN Number of top words to display
     */
    public static void analyzeLogFile(String filePath, int topN) {
        System.out.println("Analyzing file: " + filePath);
        
        // Step 1: Read the file
        String text = readTextFile(filePath);
        if (text == null) {
            return;
        }
        
        // Step 2: Extract words from the text
        List<String> words = extractWords(text);
        System.out.println("Total words found: " + words.size());
        
        // Step 3: Count word frequency
        Map<String, Integer> wordCount = countWordFrequency(words);
        
        // Step 4: Display results
        displayResults(wordCount, topN);
    }
}

File: word_frequency_analyzer.py
"""
Word Frequency Analyzer
This script reads a text file and counts the frequency of each word.
"""

import re
from collections import Counter
from pathlib import Path


def read_text_file(file_path):
    """
    Reads a text file and returns its contents as a string.
    
    Args:
        file_path (str): Path to the text file
        
    Returns:
        str: Contents of the file
    """
    try:
        with open(file_path, 'r', encoding='utf-8') as file:
            content = file.read()
        return content
    except FileNotFoundError:
        print(f"Error: File '{file_path}' not found.")
        return None
    except Exception as e:
        print(f"Error reading file: {e}")
        return None


def extract_words(text):
    """
    Extracts words from text, converting to lowercase and removing punctuation.
    
    Args:
        text (str): Input text string
        
    Returns:
        list: List of extracted words
    """
    # Use regex to find all words (alphanumeric sequences)
    # \w+ matches one or more word characters (letters, digits, underscore)
    # This ignores punctuation and splits on whitespace
    words = re.findall(r'\b\w+\b', text.lower())
    return words


def count_word_frequency(words):
    """
    Counts the frequency of each word using Counter from collections module.
    
    Args:
        words (list): List of words
        
    Returns:
        Counter: Dictionary-like object with word frequencies
    """
    # Counter automatically counts occurrences of each word
    word_count = Counter(words)
    return word_count


def display_results(word_count, top_n=10):
    """
    Displays the word frequency results in a formatted way.
    
    Args:
        word_count (Counter): Word frequency counter object
        top_n (int): Number of top words to display (default: 10)
    """
    print(f"\n{'='*50}")
    print(f"Total unique words: {len(word_count)}")
    print(f"{'='*50}\n")
    
    print(f"Top {top_n} most frequent words:")
    print(f"{'-'*50}")
    print(f"{'Word':<20} {'Frequency':<10}")
    print(f"{'-'*50}")
    
    # most_common() returns a list of tuples (word, count) sorted by frequency
    for word, count in word_count.most_common(top_n):
        print(f"{word:<20} {count:<10}")
    
    print(f"{'-'*50}\n")


def analyze_log_file(file_path, top_n=10):
    """
    Main function that orchestrates the word frequency analysis.
    
    Args:
        file_path (str): Path to the text/log file
        top_n (int): Number of top words to display
    """
    print(f"Analyzing file: {file_path}")
    
    # Step 1: Read the file
    text = read_text_file(file_path)
    if text is None:
        return
    
    # Step 2: Extract words from the text
    words = extract_words(text)
    print(f"Total words found: {len(words)}")
    
    # Step 3: Count word frequency
    word_count = count_word_frequency(words)
    
    # Step 4: Display results
    display_results(word_count, top_n)
    
    return word_count


def main():
    """
    Main entry point of the program.
    """
    # Example usage
    file_path = input("Enter the path to the text file (or press Enter for 'sample_log.txt'): ").strip()
    
    if not file_path:
        file_path = "sample_log.txt"
    
    top_n = input("Enter number of top words to display (default 10): ").strip()
    top_n = int(top_n) if top_n.isdigit() else 10
    
    # Perform the analysis
    analyze_log_file(file_path, top_n)


if __name__ == "__main__":
    main()
